### DISCUSSION TITLE
How to deploy a shared .env file

### DISCUSSION THREAD

**Discussion by viniciusbig** [deployphp/deployer#2559] [Help needed] [answered]
I'm using a cakephp recipe.
Using cake, I have a `config/.env` file with my environment variables.
But this file is not suppose to be in git (and its not), so after cloning the repo, its not there.

Since its not there, deployer doesn't do anything even the file being listed in the recipe as `sharedFile`

How can I first deploy this file if it doesn't exist?

Am I missing something?

**Comment by antonmedv** [comment]
Can you provide an example?

**Reply by viniciusbig** [reply]
I can try!

I'm deployng a fresh CakePHP 4 install for the first time.

Cake uses 3 files for configs `config/.env`, `config/app.php` and `config/app_local.php`
(only  `config/app.php` is on source control)

Cakephp recipe already says that `config/.env` and `config/app.php` in the shared_files
https://github.com/deployphp/deployer/blob/8b3d7f726c38fcf349266cb7b59b645666683949/recipe/cakephp.php#L19

But, since the file is not there, no symlink is created by deployer


---

I've just find a similar situation in Laravel
https://github.com/lorisleiva/laravel-deployer

and the text suggest that I need to run deployer the first time, create my files manually and them run it again to make this work

Is this the case?

**Comment by viniciusbig** [comment, marked as answer]
So, to give anyone else an answer here and close this question, a few thing I've found on this.

## tl;dr
For Cake4 deploy
 - Add `config/app_local.php` to your `shared_files`
 - Before first deploy, manually creates the folders `shared/config` and put your `config/app_local.php` and your `config/.env` if you need it.

---

## Deployer and Cake versions
 - Deployer 7 is still beta right now.
 - This means a fresh common install will get you Deployer 6.
 - Deployer 6 have a CakePHP3 recipe, not CakePHP4 (be aware of that)

## Config files
 - On cake4, there is 3 different files to handle environment variables: `config/.env`, `config/app.php`, `config/app_local.php`
 - `config/.env`, `config/app_local.php` should not be committed, each environment need a copy of them with the correct values.
 - So, they are *NOT* in source code control.

### So, talking about `config/.env`:
 - this file is present in Cake4 Deployer recipe as a `shared_file`; (maybe not a good default to be there)
 - the file is *NOT THERE* when Deployer clone it, and therefore, is NOT shared in the shared_folder.
 - you *should not* being using this file, check details below.
 - if you need it, you should add a new step, after `deploy:update_code` to create it (or copy from somewhere else)
 - or you can copy into `shared/config` folder

```php
task('deploy:env', function() {
    // add some code to create or copy your .env to the shared_folder
});
after('deploy:update_code', 'deploy:env');
```

### About `config/app_local.php`:

 - this is the file Cake4 uses to hold environment specific configs
 - the file is NOT THERE when cloning the repo
 - Cake's Install creates one after getting composer dependencies, but at this point Deployer already executed `deploy:shared`, so it's too late.
 - So you will need to manually copy the file into `shared/config` folder
 - And you will need to add `config/app_local.php` in your `shared_files`

```
set('shared_files', [
    'config/app_local.php',
]);
```

### Conclusion
 - You should manually create the `shared` folder, and put the `config/app_local.php` file there with the correct value
 - So now you can deploy a cakephp correctly.

### About .env files

`.env` files should not be used on production. They are not made for that. 
Having this file in production is considered a **SECURITY RISK** and also decreases the boostrap performance of your application.
